### PR TITLE
Fix HTTP/2 extended connect hang

### LINKED ISSUE #80056: Hang when HTTP/2 WebSocket connects to a HTTP/1.1 non-TLS kestrel endpoint
### Description

When using HTTP/2 without TLS it's a common mistake to connect to a HTTP/1.1 endpoint. Kestrel added special handling for this scenario by responding with a HTTP/2 GoAway HTTP_1_1_REQUIRED. Normally this works with HttpClient, but in the new HTTP/2 WebSocket scenario it hangs and times out instead.

From the stack traces, the issue appears to be here. It's stuck waiting for a settings frame that was never received.
https://github.com/dotnet/runtime/blob/c84d95d846b7cb4144904fde5f09855251c10a70/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1068-L1077

InitialSettingsReceived needs to complete/fail if the connection is aborted.
https://github.com/dotnet/runtime/blob/c84d95d846b7cb4144904fde5f09855251c10a70/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L500-L504
https://github.com/dotnet/runtime/blob/c84d95d846b7cb4144904fde5f09855251c10a70/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L518
https://github.com/dotnet/runtime/blob/c84d95d846b7cb4144904fde5f09855251c10a70/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L603

### Reproduction Steps

```c#
using System.Net;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();
app.Urls.Add("http://localhost:5000");
app.MapGet("/", () => "Hello World!");
app.Start();

using var client = new HttpClient();
var request = new HttpRequestMessage(HttpMethod.Connect, "http://localhost:5000");
request.Headers.Protocol = "websocket"; // Trigger, comment out to avoid issue
request.Version = HttpVersion.Version20;
request.VersionPolicy = HttpVersionPolicy.RequestVersionExact;
using var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
Console.WriteLine(response);
Console.WriteLine(await response.Content.ReadAsStringAsync());
```

### Expected behavior

We should get back the following error. This is what you get if you comment out the line `request.Headers.Protocol = "websocket";`
```
Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: An HTTP/2 connection could not be established because the server did not complete the HTTP/2 handshake.
 ---> System.Net.Http.HttpProtocolException: The HTTP/2 server closed the connection. HTTP/2 error code 'HTTP_1_1_REQUIRED' (0xd).
   at System.Net.Http.Http2Connection.ThrowProtocolError(Http2ProtocolErrorCode errorCode, String message)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Program.<Main>$(String[] args) in C:\temp\WebApplication126\WebApplication126\Program.cs:line 14
   at Program.<Main>(String[] args)
```

### Actual behavior

It hangs and eventually times out
```
Unhandled exception. System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.
 ---> System.TimeoutException: A task was canceled.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Program.<Main>$(String[] args) in C:\temp\WebApplication126\WebApplication126\Program.cs:line 17
   at Program.<Main>(String[] args)
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 7

### Other information

The scenario isn't supposed to succeed, but now it fails in a way that's hard to debug.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #80056

While I was updating these tests, I extended them to cover scenarios where we set different `exception.Data` entries.

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs [hunk 1]
```diff
@@ -500,10 +500,14 @@ private async Task ProcessIncomingFramesAsync()
...
+
+                    Debug.Assert(InitialSettingsReceived.Task.IsCompleted);
                 }
-                catch (IOException e)
+                catch (Exception e)
                 {
-                    throw new IOException(SR.net_http_http2_connection_not_established, e);
+                    e = new IOException(SR.net_http_http2_connection_not_established, e);
+                    InitialSettingsReceived.TrySetException(e);
+                    throw e;
```
Comment (Tratcher) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: The same exception instance shouldn't be used twice, they're not thread safe. The `throw` and the `TrySetException` need to be passed different instances.

**Review by Tratcher — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: Which part do you mean is not thread-safe? We share exceptions like this all the time.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: Given the right (wrong) interleaving of concurrent use with ExceptionDispathInfo.Capture/Throws (as is used by async/await under the covers), you can end up with call stacks that have sections from different logical flows of execution and in orders that don't make sense.  We've generally accepted that.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: > you can end up with call stacks that have sections from different logical flows of execution

Does that apply in this case given that `TrySetException` will do the capture before we get to the throw?

Comment (Tratcher) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: We had a similar issue in Kestrel, passing the same exception instance to multiple TCS's. We ended up cloning or wrapping the exceptions so each TCS got a unique instance.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: My understanding is that in cases like this where we do it in series (`TrySetException` first, then throw) it shouldn't be an issue as the `TrySetException` will do the capture before we get to the throw. Is that not the case?

We have some other cases where we'll share the same exception like [here](https://github.com/dotnet/runtime/blob/ffb3d8d49c4d3d71ec4a0a7b595e0d1a233a0a74/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1077-L1080), where that exception may be thrown from multiple threads at the same time. I take it that would be more problematic?

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: And that is only a problem for the top-most exception?
That is, an inner exception can be safely shared as long as the outer wrapper instance is different?

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: An inner exception is just arbitrary additional data reference by a wrapping Exception; it's not read or written in any way as part of throws/catches for another exception that happens to be storing a reference to it.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]: Okay, thanks for the clarifications! I'll update the code.

Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:510]:
```suggestion
                    InitialSettingsReceived.TrySetException(new IOException(SR.net_http_http2_connection_not_established, e));
                    throw new IOException(SR.net_http_http2_connection_not_established, e);
```

**Pull request merged**